lms/static/scripts/frontend_apps/components/SubmitGradeForm.tsx

```diff
@@ -95,9 +99,13 @@ export default function SubmitGradeForm({
   const [draftGradeValue, setDraftGradeValue] = useState<string | null>(null);
 
   // Track if current grade has changed compared to what was originally loaded
-  const hasUnsavedChanges = useMemo(
-    () => draftGradeValue !== null && draftGradeValue !== grade.data,
-    [draftGradeValue, grade.data]
+  const [hasUnsavedChanges, setHasUnsavedChanges] = useState<boolean>(false);
+  const updateHasUnsavedChanges = useCallback(
+    (hasUnsavedChanges: boolean) => {
+      setHasUnsavedChanges(hasUnsavedChanges);
+      onUnsavedChanges?.(hasUnsavedChanges);
+    },
+    [onUnsavedChanges]
```

Comment:
Instead of deriving `hasUnsavedChanges` from other local states, we explicitly handle it as a regular piece of state which is called every time there's an input on the grade field (and in future also the comment field), and we set it back to `false` after a successful grade submit.

This approach plays nicer with the new `onUnsavedChanges`, allowing us to invoke both the local state setter and this callback at the same time, instead of having to "listen" for changes via `useEffect` (see [Rob's comment](https://github.com/hypothesis/lms/pull/5686#discussion_r1319834066)).